R/validate.R

```diff
@@ -0,0 +1,141 @@
+#' Validate user-provided wastewater concentration data
+#'
+#' @param ww_data tibble containing the input wastewater data
+#' @param conc_col_name string indicating the name of the column containing
+#' the concentration measurements in the wastewater data
+#' @param lod_col_name string indicating the name of the column containing
+#' the limit of detection for each measurement in the wastewater data
+#' @param call Calling environment to be passed to [cli::cli_abort()] for
+#' traceback.
+#'
+#' @return NULL, invisibly
+validate_ww_conc_data <- function(ww_data,
+                                  conc_col_name,
+                                  lod_col_name,
+                                  call = rlang::caller_env()) {
+  ww_conc <- ww_data |> dplyr::pull({
+    conc_col_name
+  })
+  arg <- conc_col_name
+  assert_non_missingness(ww_conc, arg, call)
+  assert_elements_non_neg(ww_conc, arg, call,
+    add_err_msg = c(
+      "Note that the model expects natural ",
+      "scale concentration values, ",
+      "which must be non-negative"
+    )
+  )
+  checkmate::assert_vector(ww_conc)
+
+  ww_lod <- ww_data |> dplyr::pull({
+    lod_col_name
+  })
+  arg <- "lod_col_name"
+  assert_non_missingness(ww_lod, arg, )
+  assert_elements_non_neg(ww_lod, arg, call,
+    add_err_msg = c(
+      "Note that the model expects natural ",
+      "scale LOD values, which must be",
+      "non-negative"
+    )
+  )
+  checkmate::assert_vector(ww_lod)
+
+  # Wastewater date column should be of date type!
+  ww_obs_dates <- ww_data$date
+  arg <- "ww_obs_dates"
+  assert_non_missingness(ww_obs_dates, arg, call)
+  checkmate::assert_date(ww_obs_dates)
+  # check for multiple observations per day within a site and lab
+  ww_data |>
+    dplyr::group_by(site, lab) |>
+    assert_no_repeated_elements()
```

Comment:
I am fairly sure this will not work as intended. `assert_no_repeated_elements()` expects a vector, not a data frame. `duplicated()` will work on a data frame or `tibble`, but it identifies cases where _all columns are duplicated_, which is not what you want here, and it also will not automatically be applied over `dplyr::group_by()` groups. Example:

```R
> df <- tibble::tibble(x=c(1,1,1,2,2), y=c(1,1:4), z=c(1,1,1,1,1))
> df
# A tibble: 5 × 3
      x     y     z
  <dbl> <dbl> <dbl>
1     1     1     1
2     1     1     1
3     1     2     1
4     2     3     1
5     2     4     1
>
> df |> duplicated()
[1] FALSE  TRUE FALSE FALSE FALSE
>
> df |> dplyr::group_by(x,y) |> duplicated()
[1] FALSE  TRUE FALSE FALSE FALSE
>
> df |> dplyr::group_by(x,y) |> dplyr::summarise(dup = any(duplicated(z)))
`summarise()` has grouped output by 'x'. You can override using the `.groups`
argument.
# A tibble: 4 × 3
# Groups:   x [2]
      x     y dup
  <dbl> <dbl> <lgl>
1     1     1 TRUE
2     1     2 FALSE
3     2     3 FALSE
4     2     4 FALSE

Reply:
For this reason, it will be important to implement tests for these checkers, ideally checking at least one expected pass and one expected failure for each.

Reply:
Also I got confused and these should be all the same :( (so all repeats)